sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -406,6 +406,31 @@ ze_result_t ZeCall::doCall(ze_result_t ZeResult, const char *CallStr,
     return mapError(Result);
 #define ZE_CALL_NOCHECK(Call) ZeCall().doCall(Call, #Call, false)
 
+// This helper function creates a pi_event and associate a pi_queue.
+// We need to increment the reference counter here to avoid pi_queue
+// being release before the associated pi_event is released because
+// piEventRelease requires access to the associated pi_queue.
+// In piEventRelease, the reference counter is decremented to release it.
+inline pi_result createEventAndAssociateQueue(
+    pi_queue Queue, pi_event *Event, pi_command_type CommandType,
+    ze_command_list_handle_t ZeCommandList, ze_event_handle_t *ZeEvent,
+    bool QueueLockAcquired) {
+  pi_result Res = piEventCreate(Queue->Context, Event);
+  if (Res != PI_SUCCESS)
+    return Res;
+
+  (*Event)->Queue = Queue;
+  (*Event)->CommandType = CommandType;
+  (*Event)->ZeCommandList = ZeCommandList;
+
+  *ZeEvent = (*Event)->ZeEvent;
+  if (QueueLockAcquired)
+    Queue->RefCount++;
```

Comment:
Please outline everything but the queue locking from piQueueRetain into piQueueRetainNoLock and call piQueueRetainNoLock instead here. It will be easier to conduct retain logic changes.

Reply:
Yes, I removed the LockAcquired condition entirely from this function.
Do you mean this?
`void piQueueRetainNoLock(pi_queue Queue) { Queue->RefCount++; }

...
*ZeEvent = (*Event)->ZeEvent;
piQueueRetainNoLock(Queue);
....
'

Reply:
yes. And also use piQueueRetainNoLock in the piQueueRetain.

Reply:
Got it. Thanks.

Reply:
Done. Thanks.